feat(docs-infra): add shortcuts for the angular.io PWA

gkalpak opened the pull request:
This commits adds some shortcut definitions for the angular.io PWA. The user agent can use them to assemble a context menu to be displayed by the operating system when a user engages with the app's icon. (In addition, shortcuts provide an easy way for users to add links to specific pages on their home screen.)

See [here][1] for more details on the `shortcuts` property of the PWA manifest.

The choice of pages to create shortcuts to was influenced by the following facts/criteria:
- It seems that only the first 4 shortcuts are displayed by Chrome (at least on my Android phone).
- Since the PWA is mostly used on mobile, I omitted pages that are less likely to be useful for mobile users (such as pages related to CLI).

##
**TODO**

- [x] Decide what icons to use (if any). (See https://github.com/angular/angular/pull/40393#issuecomment-758189046.)

[1]: https://developer.mozilla.org/en-US/docs/Web/Manifest/shortcuts

mary-poppins commented:
You can preview 64bab94 at https://pr40393-64bab94.ngbuilds.io/.

gkalpak commented:
A couple of images to get an idea of how this looks on my phone:

_**Context menu with shortcuts for the angular.io PWA:**_
<img
    src="https://user-images.githubusercontent.com/8604205/104230162-e1ab8600-5455-11eb-8bdf-00d8503ad0db.png"
    alt="Context menu with shortcuts for the angular.io PWA"
    width="350" />

_**A direct link to the API reference page created using the corresponding shortcut:**_
<img
    src="https://user-images.githubusercontent.com/8604205/104230181-e839fd80-5455-11eb-8803-34eb99ad3791.png"
    alt="Direct link to API page (created via API page shortcut)"
    width="350" />

##
**Note on Icons**

I've picked some icons from [concept-icons][1] that seemed appropriate-ish, but they do in fact represent different things. We should decide what to do with icons (which icons to use or not use icons at all).

[1]: https://github.com/angular/angular/tree/76954688b18f1509d4c9cca41b8e113ee423d4db/aio/content/images/marketing/concept-icons

IgorMinar reviewed (approved):
LGTM! Thank you

IgorMinar commented:
let's drop the icons to decrease the risk of confusion since they are already used for different concepts.

gkalpak committed:
feat(docs-infra): add shortcuts for the angular.io PWA

This commits adds some shortcut definitions for the angular.io PWA. The
user agent can use them to assemble a context menu to be displayed by
the operating system when a user engages with the app's icon. (In
addition, shortcuts provide an easy way for users to add links to
specific pages on their home screen.)

See [here][1] for more details on the `shortcuts` property of the PWA
manifest.

The choice of pages to create shortcuts to was influenced by the
following facts/criteria:
- It seems that only the first 4 shortcuts are displayed by Chrome (at
  least on my Android phone).
- Since the PWA is mostly used on mobile, I omitted pages that are less
  likely to be useful for mobile users (such as pages related to CLI).

[1]: https://developer.mozilla.org/en-US/docs/Web/Manifest/shortcuts

gkalpak committed:
refactor(docs-infra): more some `pwa-manifest.json` keys to the top to improve readability

This commit moves some properties with simple, string values in
`pwa-manifest.json` to the top (i.e. above the more lengthy `icons`
array) to make them easier to discover.

gkalpak committed:
fixup! feat(docs-infra): add shortcuts for the angular.io PWA

mary-poppins commented:
You can preview bd30ae6 at https://pr40393-bd30ae6.ngbuilds.io/.

mary-poppins commented:
You can preview 37cd545 at https://pr40393-37cd545.ngbuilds.io/.

gkalpak commented:
I removed the icons and added a test to ensure the shortcut URLs exist (as discussed).

petebacondarwin commented on aio/tests/e2e/src/pwa-manifest.e2e-spec.ts:
```diff
@@ -0,0 +1,29 @@
+      shortcuts = await page.getPwaShortcuts();
+    });
+
+    it('should exist', async () => {
+      for (const {short_name, url} of shortcuts) {
```
```suggestion
    for (const {short_name, url} of shortcuts) {
      it(`${short_name} should exist`, async () => {
```

gkalpak replied:
I would be prefer that too, but that is not possible unfortunately, because `shortcuts` are fetched asynchronously (so they are not available until we are inside the `it()` function.

The only other way to make it work that I know of is to fetch the shortcuts in Protractor's `onPrepare()` hook (as we do for the [deployment e2e tests][1]), but I felt this was adding a level of indirection that was not worth it here (given that we are going to have very few shortcuts).

[1]: https://github.com/angular/angular/blob/9a5ac47331faf0f6030bff75ad8ef8dab2f8d2d9/aio/tests/deployment/e2e/protractor.conf.js#L41-L57

gkalpak replied:
Of course, if you have ideas on how to get this to work in a straight forward way, I'm all ears :ear: :grin:

petebacondarwin replied:
Blimey - of course I forgot that `shortcuts` is pulled in async. Nevermind.

petebacondarwin commented on aio/tests/e2e/src/pwa-manifest.po.ts:
```diff
@@ -0,0 +1,74 @@
+ * For simplicity, we only define types for the properties we care about in tests.
+ * @see https://developer.mozilla.org/en-US/docs/Web/Manifest
+ */
+export type PwaManifest = Json & {
+  shortcuts?: PwaShortcutItem[];
+};
```
I believe the norm is that if you are describing a type literal then the delimiter is commas:

```suggestion
export type PwaManifest = Json & {
  shortcuts?: PwaShortcutItem[],
};
```

(and similarly for `PwaShortcutItem` below).

gkalpak replied:
Done.

petebacondarwin replied:
I turns out that there is no "norm" - not even in our own codebase here. So I am happy for you to stick with semi-colons as you wish.

gkalpak replied:
Too late 😛

gkalpak committed:
fixup! feat(docs-infra): add shortcuts for the angular.io PWA

mary-poppins commented:
You can preview e35398f at https://pr40393-e35398f.ngbuilds.io/.
